// @wdio/static-server-service/build/launcher.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _express = _interopRequireDefault(require("express"));

var _fsExtra = _interopRequireDefault(require("fs-extra"));

var _morgan = _interopRequireDefault(require("morgan"));

var _path = _interopRequireDefault(require("path"));

var _logger = _interopRequireDefault(require("@wdio/logger"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const log = (0, _logger.default)('@wdio/static-server-service');
const DEFAULT_LOG_NAME = 'wdio-static-server-service.log';

class StaticServerLauncher {
  onPrepare({
    outputDir,
    staticServerFolders: folders,
    staticServerPort: port = 4567,
    staticServerMiddleware: middleware = []
  }) {
    if (!folders) {
      return;
    }

    this.server = (0, _express.default)();
    this.folders = folders;
    this.port = port;

    if (outputDir) {
      const file = _path.default.join(outputDir, DEFAULT_LOG_NAME);

      _fsExtra.default.createFileSync(file);

      const stream = _fsExtra.default.createWriteStream(file);

      this.server.use((0, _morgan.default)('tiny', {
        stream
      }));
    }

    (Array.isArray(folders) ? folders : [folders]).forEach(folder => {
      log.info('Mounting folder `%s` at `%s`', _path.default.resolve(folder.path), folder.mount);
      this.server.use(folder.mount, _express.default.static(folder.path));
    });
    middleware.forEach(ware => this.server.use(ware.mount, ware.middleware));
    return new Promise((resolve, reject) => this.server.listen(this.port, err => {
      if (err) {
        log.error(`Couldn't start static server: ${err.message}`);
        return reject(err);
      }

      log.info(`Static server running at http://localhost:${port}`);
      resolve();
    }));
  }

}

exports.default = StaticServerLauncher;
